examples/apigateway/main.go

```diff
@@ -0,0 +1,203 @@
+	log "github.com/go-kit/kit/log"
+	//grpctransport "github.com/go-kit/kit/transport/grpc"
+	httptransport "github.com/go-kit/kit/transport/http"
+	//proto "github.com/golang/protobuf/proto"
+	"github.com/gorilla/mux"
+	"github.com/hashicorp/consul/api"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+func main() {
+	fs := flag.NewFlagSet("", flag.ExitOnError)
+	var (
+		httpAddr   = fs.String("http.addr", ":8000", "Address for HTTP (JSON) server")
+		consulAddr = fs.String("consul.addr", "", "Consul agent address")
+	)
+	flag.Usage = fs.Usage
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		fmt.Fprintf(os.Stderr, "%v", err)
+		os.Exit(1)
+	}
+
+	// log
+	logger := log.NewLogfmtLogger(os.Stderr)
+	logger = log.NewContext(logger).With("ts", log.DefaultTimestampUTC).With("caller", log.DefaultCaller)
+	stdlog.SetFlags(0)                             // flags are handled by Go kit's logger
+	stdlog.SetOutput(log.NewStdlibAdapter(logger)) // redirect anything using stdlib log to us
+
+	// errors
+	errc := make(chan error)
+	go func() {
+		errc <- interrupt()
+	}()
+
+	// consul
+	consulConfig := api.DefaultConfig()
+	if len(*consulAddr) > 0 {
+		consulConfig.Address = *consulAddr
+	}
+	consulClient, err := api.NewClient(consulConfig)
+	if err != nil {
+		logger.Log("fatal", err)
+	}
+	discoveryClient := consul.NewClient(consulClient)
+
+	ctx := context.Background()
+
+	// service definitions
+	serviceDefs := []*ServiceDef{}
+	serviceDefs = append(serviceDefs, &ServiceDef{
+		Name: "addsvc",
+		Endpoints: map[string]loadbalancer.Factory{
+			"/api/addsvc/concat": factoryAddsvc(ctx, logger, makeConcatEndpoint),
+			"/api/addsvc/sum":    factoryAddsvc(ctx, logger, makeSumEndpoint),
+		},
+	})
+	serviceDefs = append(serviceDefs, &ServiceDef{
+		Name: "stringsvc",
+		Endpoints: map[string]loadbalancer.Factory{
+			"/api/stringsvc/uppercase": routeFactory(ctx, "uppercase"),
+			"/api/stringsvc/count":     routeFactory(ctx, "count"),
+		},
+	})
+
+	// discover instances and register endpoints
+	r := mux.NewRouter()
+	for _, def := range serviceDefs {
+		for path, e := range def.Endpoints {
+			pub, err := consul.NewPublisher(discoveryClient, e, logger, def.Name)
+			if err != nil {
+				logger.Log("fatal", err)
+			}
+			r.HandleFunc(path, makeHandler(ctx, loadbalancer.NewRoundRobin(pub), logger))
+		}
+	}
+
+	// apigateway
+	go func() {
+		errc <- http.ListenAndServe(*httpAddr, r)
+	}()
+
+	// wait for interrupt/error
+	logger.Log("fatal", <-errc)
+}
+
+type ServiceDef struct {
+	Name      string
+	Endpoints map[string]loadbalancer.Factory
+}
+
+func interrupt() error {
+	c := make(chan os.Signal)
+	signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
+	return fmt.Errorf("%s", <-c)
+}
+
+func makeHandler(ctx context.Context, lb loadbalancer.LoadBalancer, logger log.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		e, err := lb.Endpoint()
+		if err != nil {
+			logger.Log("error", err)
+			return
```

Comment:
I think you have to write something to the ResponseWriter with `http.Error` or something..